pkg/helmdeploy/write.go

```diff
@@ -0,0 +1,212 @@
+		}
+
+		archive, err := base.FindHelmChartArchiveInRelease(u.Files, helmChart)
+		if err != nil {
+			return errors.Wrap(err, "failed to find helm chart archive in release")
+		}
+
+		archivePath := path.Join(chartDir, fmt.Sprintf("%s-%s.tgz", helmChart.Spec.Chart.Name, helmChart.Spec.Chart.ChartVersion))
+		if err := ioutil.WriteFile(archivePath, archive, 0644); err != nil {
+			return errors.Wrap(err, "failed to write helm chart archive")
+		}
+
+		mergedValues := helmChart.Spec.Values
+		if mergedValues == nil {
+			mergedValues = map[string]kotsv1beta2.MappedChartValue{}
+		}
+		for _, optionalValues := range helmChart.Spec.OptionalValues {
+			parsedBool, err := strconv.ParseBool(optionalValues.When)
+			if err != nil {
+				return errors.Wrap(err, "failed to parse when conditional on optional values")
+			}
+			if !parsedBool {
+				continue
+			}
+			if optionalValues.RecursiveMerge {
+				mergedValues = kotsv1beta2.MergeHelmChartValues(mergedValues, optionalValues.Values)
+			} else {
+				for k, v := range optionalValues.Values {
+					mergedValues[k] = v
+				}
+			}
+		}
+
+		helmValues, err := helmChart.Spec.GetHelmValues(mergedValues)
+		if err != nil {
+			return errors.Wrap(err, "failed to render local values for chart")
+		}
+
+		valuesContent, err := yaml.Marshal(helmValues)
+		if err != nil {
+			return errors.Wrap(err, "failed to marshal values")
+		}
+
+		builder, _, err := base.NewConfigContextTemplateBuilder(u, renderOptions)
+		if err != nil {
+			return errors.Wrap(err, "failed to create config context template builder")
+		}
+
+		renderedValuesContent, err := builder.RenderTemplate(fmt.Sprintf("%s-values", helmChart.GetDirName()), string(valuesContent))
+		if err != nil {
+			return errors.Wrap(err, "failed to render values")
+		}
+
+		valuesPath := path.Join(chartDir, "values.yaml")
+		if err := ioutil.WriteFile(valuesPath, []byte(renderedValuesContent), 0644); err != nil {
+			return errors.Wrap(err, "failed to write values file")
+		}
+	}
+
+	return nil
+}
+
+type WriteOptions struct {
+	HelmDir             string
+	RenderedDir         string
+	Log                 *logger.CLILogger
+	Downstreams         []string
+	KotsKinds           *kotsutil.KotsKinds
+	ProcessImageOptions midstream.ProcessImageOptions
+	Clientset           kubernetes.Interface
+}
+
+// WriteRenderedHelmCharts writes the rendered helm chart to the rendered directory
+func WriteRenderedHelmCharts(opts WriteOptions) error {
+	if opts.KotsKinds == nil || opts.KotsKinds.V1Beta2HelmCharts == nil {
+		return nil
+	}
+
+	for _, downstream := range opts.Downstreams {
+		for _, helmChart := range opts.KotsKinds.V1Beta2HelmCharts.Items {
+			renderedPath, err := renderHelmChart(opts, downstream, &helmChart)
+			if err != nil {
+				return errors.Wrap(err, "failed to render helm chart")
+			}
+
+			if err := processImages(opts, renderedPath); err != nil {
+				return errors.Wrap(err, "failed to process images")
+			}
```

Comment:
This is the first place where we have the rendered chart manifests, so we push images here.  Is there a better place to do this, or could this cause any problems?  This would normally be done as part of the midstream flow, but that doesn't exist in the new helm deploy.